File: frontend/src/types/market.ts
export interface OrderBookLevel {
  price: number
  quantity: number
}

export interface UnifiedOrderBook {
  msg_id: string
  exchange: string
  symbol: string
  timestamp: number
  sequence: number
  bids: OrderBookLevel[]
  asks: OrderBookLevel[]
}

export interface UnifiedTrade {
  msg_id: string
  exchange: string
  symbol: string
  trade_id: string
  price: number
  quantity: number
  side: 'buy' | 'sell'
  timestamp: number
}

export interface SpreadSnapshot {
  symbol: string
  exchange_a: string
  exchange_b: string
  a_best_bid: number
  a_best_ask: number
  b_best_bid: number
  b_best_ask: number
  spread: number
  spread_pct: number
  timestamp: number
}

export interface SpreadCell {
  exchange_a: string
  exchange_b: string
  spread_ab: number
  spread_ba: number
  best_spread: number
  spread_pct: number
}

export interface SpreadMatrix {
  symbol: string
  exchanges: string[]
  cells: SpreadCell[]
  stale_exchanges: string[]
  timestamp: number
}

export interface SpreadAlertEvent {
  symbol: string
  exchange_a: string
  exchange_b: string
  spread_pct: number
  direction: string
  timestamp: number
}

export interface WSMessage {
  type: 'orderbook' | 'trade' | 'spread' | 'spread_matrix' | 'spread_alert'
  data: UnifiedOrderBook | UnifiedTrade | SpreadSnapshot | SpreadMatrix | SpreadAlertEvent
}

File: backend/app/models/market.py
from pydantic import BaseModel
from typing import Literal
from decimal import Decimal


class OrderBookLevel(BaseModel):
    price: Decimal
    quantity: Decimal


class UnifiedOrderBook(BaseModel):
    msg_id: str
    exchange: str
    symbol: str
    timestamp: int
    sequence: int
    bids: list[OrderBookLevel]
    asks: list[OrderBookLevel]


class UnifiedTrade(BaseModel):
    msg_id: str
    exchange: str
    symbol: str
    trade_id: str
    price: Decimal
    quantity: Decimal
    side: Literal["buy", "sell"]
    timestamp: int


class SpreadSnapshot(BaseModel):
    symbol: str
    exchange_a: str
    exchange_b: str
    a_best_bid: Decimal
    a_best_ask: Decimal
    b_best_bid: Decimal
    b_best_ask: Decimal
    spread: Decimal
    spread_pct: Decimal
    timestamp: int


class SpreadCell(BaseModel):
    exchange_a: str
    exchange_b: str
    spread_ab: Decimal
    spread_ba: Decimal
    best_spread: Decimal
    spread_pct: Decimal


class SpreadMatrix(BaseModel):
    symbol: str
    exchanges: list[str]
    cells: list[SpreadCell]
    stale_exchanges: list[str]
    timestamp: int


class SpreadAlertEvent(BaseModel):
    symbol: str
    exchange_a: str
    exchange_b: str
    spread_pct: Decimal
    direction: str
    timestamp: int
